### PR TITLE
Version 2

### PR DESCRIPTION
**Pull request by HolyWalley**
Full Refactored

### CODE REVIEW

**Comment by HolyWalley** [comment]
@anatoliliotych check it.

File: CurrentUser.rb [hunk 1]
```diff
@@ -0,0 +1,23 @@
+module CurrentUser
+  def auth?
+    token = session[:token]
+    return false if token.nil?
+    begin
+      payload, header = JWT.decode(token, settings.verify_key, true)
+    rescue JWT::DecodeError => e
+      puts e.inspect
```
Comment (anatoliliotych) [inline CurrentUser.rb]: use logger for such things.

File: CurrentUser.rb [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    end
+    return false if payload.nil? || header.nil?
+    exp = header['exp']
```
Comment (anatoliliotych) [inline CurrentUser.rb]: merge this and next line

File: CurrentUser.rb [hunk 3]
```diff
@@ -0,0 +1,23 @@
+module CurrentUser
+  def auth?
+    token = session[:token]
+    return false if token.nil?
```
Comment (anatoliliotych) [inline CurrentUser.rb]: do we need additional nil check?

File: app.rb [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+
+def add_user(user_id, token, exp)
+  user = User.find_by(uid: user_id)
+
+  exp = (Time.now.to_i + exp).to_s
+
+  headers = { exp: exp }
+  session[:token] = JWT.encode({ user_id: user_id }, settings.signing_key, 'RS256', headers)
+
+  User.create(uid: user_id, token: token, token_dying_time: exp) if user.nil?
```
Comment (anatoliliotych) [inline app.rb]: no need to check for nil

File: app.rb [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+set :session_secret, SESSION_SECRET
+
+include CurrentUser
+
+get '/' do
+  erb :index
+end
+
+def add_user(user_id, token, exp)
+  user = User.find_by(uid: user_id)
```
Comment (anatoliliotych) [inline app.rb]: you can use find_or_create and avoid creation later

File: app.rb [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+  User.create(uid: user_id, token: token, token_dying_time: exp) if user.nil?
+  User.update(token: token, token_dying_time: exp)
+  redirect to('/feed')
+end
+
+get '/callback' do
+  redirect to('/') unless code = params[:code]
+
+  uri = URI("#{OAUTH_VK}access_token?#{CLIENT_ID}&#{VK_SECRET}&redirect_uri=#{CALLBACK}&code=#{code}")
+  vk_response = Net::HTTP.get(uri)
```
Comment (anatoliliotych) [inline app.rb]: try to name differently because you reuse the same var

File: app.rb [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+  vk_response = JSON.parse(vk_response)
+  redirect to('/') unless vk_response['error'].nil?
```
Comment (anatoliliotych) [inline app.rb]: do we need nil?

File: app.rb [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+end
+
+def get_user_friends_photos
+  user_id = current_user
+  token = User.find_by(uid: user_id)['token']
+  photos = {}
+  Photo.where(uid: user_id).each do |elem|
+    i = 0
+    while i < elem['photocount'].to_i
+      uri = URI("#{VK_API_URL}photos.get?user_id=#{elem.fid}&album_id=saved&count=#{MAX_PHOTO}&offset=#{i}&v=5.52&access_token=#{token}")
```
Comment (anatoliliotych) [inline app.rb]: move this url preparation to method

File: app.rb [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+        photos[photo['date']] = photo['photo_604']
+      end
+      i += MAX_PHOTO
+    end
+  end
+  photos.sort.reverse
+end
+
+get '/feed' do
+  @photos = {}
```
Comment (anatoliliotych) [inline app.rb]: no need in this line if method get_user_frends_photos returns Hash in any case

File: app.rb [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+  @photos = get_user_friends_photos
+  erb :feed
+end
+
+get '/add' do
+  @s = get_added_list
```
Comment (anatoliliotych) [inline app.rb]: what is @s??? please use proper naming

File: app.rb [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+  erb :add
+end
+
+def get_added_list
+  user_id = current_user
+  token = User.find_by(uid: user_id)['token']
+  s = ''
```
Comment (anatoliliotych) [inline app.rb]: what is s?

File: app.rb [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+    user_hash = JSON.parse(user_json)
+    s += "<span><a href = 'https://vk.com/id#{elem.fid}'>" +
+         user_hash['response'][0]['first_name'] + ' ' +
+         user_hash['response'][0]['last_name'] +
+         "</a> <a class = \'dell\'href = \'#\'><img src = \'img/icons8-Delete-64.png\' class=\'cross\' alt = \'#{elem.fid}\'></a><br></span>"
+  end
+  s
+end
+
+def get_fid(f_name, user_token, user_id)
```
Comment (anatoliliotych) [inline app.rb]: what is fid?

File: app.rb [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+  redirect to('/add') unless user_hash['error'].nil?
+  fid = user_hash['response'][0]['id']
+  redirect to('/feed') if Photo.find_by(fid: fid, uid: user_id)
+  fid
+end
+
+post '/add' do
+  user_id = current_user
+  token = User.find_by(uid: user_id)['token']
+  f_name = params['lastphoto']['fid']
```
Comment (anatoliliotych) [inline app.rb]: what is f_name?

File: app.rb [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+  uri = URI("#{VK_API_URL}photos.get?user_id=#{fid}&album_id=saved&count=1&v=5.52&access_token=#{token}")
+  photos_json = Net::HTTP.get(uri)
+  photos_hash = JSON.parse(photos_json)
+
+  redirect to('/closed') unless photos_hash['error'].nil?
+  Photo.create(uid: user_id, fid: fid, photocount: photos_hash['response']['count'])
+  redirect to('/add')
+end
+
+post '/dell' do
```
Comment (anatoliliotych) [inline app.rb]: dell is strange word

File: app.rb [hunk 15]
```diff
@@ -0,0 +1,174 @@
...
+  Photo.find_by(fid: fid, uid: user_id).destroy
+end
+
+get '/logout' do
+  session[:token] = nil
+  redirect to('/')
+end
+
+post '/upload' do
+  @photos = {}
```
Comment (anatoliliotych) [inline app.rb]: remove this

File: app.rb [hunk 16]
```diff
@@ -0,0 +1,174 @@
...
+  fid = params[:fid]
+  Photo.find_by(fid: fid, uid: user_id).destroy
+end
+
+get '/logout' do
+  session[:token] = nil
+  redirect to('/')
+end
+
+post '/upload' do
```
Comment (anatoliliotych) [inline app.rb]: rewrite this with proper naming...

File: views/add.erb [hunk 17]
```diff
@@ -0,0 +1,21 @@
+<div id = 'results'>
+	<div class = 'content'>
```
Comment (anatoliliotych) [inline views/add.erb]: what is wrong with idents?

Comment (HolyWalley) [inline views/add.erb]: ??

**Comment by HolyWalley** [comment]
@anatoliliotych done

File: views/layout.erb [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+</head>
+<body>
+
+  <div class = "menuContainer">
+    <ul>
+      <li><a href="/feed" class="feed">feed</a></li>
+      <li><a href="/add" class="friends">friends</a></li>
+      <li><a href="https://twitter.com/doowalley" class="thanks">thanks</a></li>
+      <% if auth? %>
+      <li><a href="/logout" class="logout">logout</a></li>
```
Comment (anatoliliotych) [inline views/layout.erb]: ident is wrong

File: views/layout.erb [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+  <meta charset="utf-8">
+
+  <title>VK SAVED PHOTO FEED</title>
+  <link rel="stylesheet" type="text/css" href="inputbox.css">
+  <link rel="stylesheet" type="text/css" href="styles.css">
+  <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.2.1/jquery.min.js"></script>
+  <script src="scripts.js"></script>
+
+</head>
+<body>
```
Comment (anatoliliotych) [inline views/layout.erb]: ident is wrong

File: views/feed.erb [hunk 20]
```diff
@@ -0,0 +1,15 @@
...
+  <img src= <%= @photos[2][1] %>>
+<% else %>
+  <p id = 'addSomeone'>
+    Hi. This service provides the opportunity to watch the tape saved photos selected people.
+    <br><a href = "/add">Add someone</a>
+  </p>
+<% end %>
+</div>
+</div>
+
```
Comment (anatoliliotych) [inline views/feed.erb]: remove line

**Comment by HolyWalley** [comment]
@anatoliliotych done

File: CurrentUser.rb [hunk 21]
```diff
@@ -0,0 +1,21 @@
+module CurrentUser
+  def auth?
+    token = session[:token]
+    return false unless token
+    begin
+      payload, header = JWT.decode(token, settings.verify_key, true)
+    rescue JWT::DecodeError => e
+    end
+    return false if payload.nil? || header.nil?
```
Comment (anatoliliotych) [inline CurrentUser.rb]: don't use .nil? check when this is required.

File: CurrentUser.rb [hunk 22]
```diff
@@ -0,0 +1,21 @@
+module CurrentUser
+  def auth?
+    token = session[:token]
+    return false unless token
+    begin
+      payload, header = JWT.decode(token, settings.verify_key, true)
+    rescue JWT::DecodeError => e
```
Comment (anatoliliotych) [inline CurrentUser.rb]: in such case you can do oneliner

File: app.rb [hunk 23]
```diff
@@ -0,0 +1,174 @@
...
+OAUTH_VK = 'https://oauth.vk.com/'.freeze
+SCOPES = 'scope=friends,photos'.freeze
+AUTHO_URL = "#{OAUTH_VK}authorize?#{CLIENT_ID}&display=page&redirect_uri=#{CALLBACK}&#{SCOPES}&response_type=code&v=5.67".freeze
+VK_API_URL = 'https://api.vk.com/method/'.freeze
+MAX_PHOTO = 1_000
+
+signing_key_path = File.expand_path('../app.rsa', __FILE__)
+verify_key_path = File.expand_path('../app.rsa.pub', __FILE__)
+
+signing_key = ''
```
Comment (anatoliliotych) [inline app.rb]: could be removed

File: app.rb [hunk 24]
```diff
@@ -0,0 +1,174 @@
...
+verify_key = ''
+
+File.open(signing_key_path) do |file|
+  signing_key = OpenSSL::PKey.read(file)
```
Comment (anatoliliotych) [inline app.rb]: could be simplified

File: app.rb [hunk 25]
```diff
@@ -0,0 +1,174 @@
...
+  erb :add
+end
+
+def get_added_list
+  user_id = current_user
+  token = User.find_by(uid: user_id)['token']
+  added_friends = ''
+  Photo.where(uid: user_id).each do |elem|
+    uri = URI("#{VK_API_URL}users.get?user_id=#{elem.fid}&v=5.52&access_token=#{token}")
+    user_json = Net::HTTP.get(uri)
```
Comment (anatoliliotych) [inline app.rb]: move to method

**Comment by HolyWalley** [comment]
@anatoliliotych check it :)

**Pull request merged**